test/gtest/conv_immediate.cpp

```diff
@@ -0,0 +1,904 @@
+ * from 2048~4096, half can only express 1/2 the number. number 2049, 2051, 2053, 2055.... can not
+ * be expressed. (max interval is 2) from 4096~8192, half can only express 1/4 the number. number
+ * 4097, 4098, 4099, 4101, 4102, 4103, 4105, 4106, 4107, 4109... can not expressd. (max interval is
+ * 4) from 8192~16384, half can only express 1/8 the number. (max interval is 8)
+ */
+template <typename T>
+void rand_tensor_integer(tensor<T>& t, int max = RAND_INTEGER_MAX, int min = RAND_INTEGER_MIN)
+{
+    // use integer to random.
+    for(size_t i = 0; i < t.data.size(); i++)
+        t[i] = static_cast<T>(prng::gen_A_to_B(min, max));
+}
+
+template <typename T>
+bool verify_tensor(tensor<T>& t_gpu,
+                   tensor<T>& t_cpu,
+                   float integer_interval = MAX_INTEGER_INTERVAL)
+{
+    if(t_gpu.data.size() != t_cpu.data.size())
+    {
+        MIOPEN_LOG_E("size not equal, should not happen");
+        return false;
+    }
+    auto idx          = miopen::mismatch_idx(t_gpu.data, t_cpu.data, miopen::float_equal);
+    bool valid_result = true;
+    if(idx < miopen::range_distance(t_cpu))
+    {
+        // give a re-try chance for half_float
+        // max gemm_k is wrw, max_n=2, max_ho/wo=14, max integer=4, max value=2*14*14*4*4 = 6272.
+        // hence max integer interval is 4
+        // for gpu we cast value to float, for cpu we cast value to double. hence inside kernel
+        // precision is guaranteed.
+        // the problem is cast the result back.
+        // round-to-nearest(default rounding mode) seems will have little difference when doing
+        // double->half, compare to float->half.
+        // hence we give a chance to calculate if the difference is still following our experience,
+        // while doing integer computation.
+        auto max_diff = miopen::max_diff(t_gpu, t_cpu);
+        if(max_diff > integer_interval)
+            valid_result = false;
+    }
+
+    if(!valid_result)
+    {
+        std::cout << "diff at:" << idx << ", gpu:" << t_gpu[idx] << ", cpu:" << t_cpu[idx]
+                  << std::endl;
+    }
+    return valid_result;
+}
+
+/// input: a vector of lengths of dims in a tensor
+/// multiply each element with a random constant integer
+void pad_tensor_strides(std::vector<int>& strides)
+{
+    constexpr int min_stride_multiplier = 1;
+    constexpr int max_stride_multiplier = 5;
+
+    auto c = prng::gen_A_to_B(min_stride_multiplier, max_stride_multiplier);
+    for(auto& v : strides)
+    {
+        // cppcheck-suppress useStlAlgorithm
+        v = v * c;
+    }
+}
+
+template <miopen::conv::Direction direction,
+          typename TREF,
+          typename TOUT,
+          miopenTensorLayout_t tensor_layout>
+struct TypeDefs
+{
+    static constexpr miopen::conv::Direction get_direction() { return direction; };
+    typedef TREF TRef;
+    typedef TOUT Tout;
+    static constexpr miopenTensorLayout_t get_tensor_layout() { return tensor_layout; };
```

Comment:
I don't think it's significant difference between `constexpr` function and static member, I just prefer static member to keep using this aggregate uniformly across the types and constants. Moreover all the `std::` constants are constants, not functions, check `std::is_same_v` implementation.
But `using` must be used instead of `typedef`. It is more straight and potentially more flexible.
Inverted typedef logic is kind of ugly.
```suggestion
    static constexpr miopen::conv::Direction Direction = direction;
    using TRef = TREF;
    using Tout = TOUT;
    static constexpr miopenTensorLayout_t Layout = tensor_layout;
```

Reply:
accepted